app/com/arpnetworking/metrics/portal/alerts/impl/DatabaseAlertRepository.java

```diff
@@ -0,0 +1,365 @@
+            if (ebeanAlert == null) {
+                ebeanAlert = new models.ebean.Alert();
+                created = true;
+            }
+
+            ebeanAlert.setCluster(alert.getCluster());
+            ebeanAlert.setUuid(alert.getId());
+            ebeanAlert.setMetric(alert.getMetric());
+            ebeanAlert.setContext(alert.getContext());
+            ebeanAlert.setNagiosExtension(getNagiosExtensionFromMap(alert.getExtensions()));
+            ebeanAlert.setName(alert.getName());
+            ebeanAlert.setOperator(alert.getOperator());
+            ebeanAlert.setPeriod(alert.getPeriod().toStandardSeconds().getSeconds());
+            ebeanAlert.setQuantityValue(alert.getValue().getValue());
+            ebeanAlert.setQuantityUnit(alert.getValue().getUnit().isPresent() ? alert.getValue().getUnit().get() : null);
+            ebeanAlert.setStatistic(alert.getStatistic());
+            ebeanAlert.setService(alert.getService());
+            _alertQueryGenerator.saveAlert(ebeanAlert);
+            transaction.commit();
+
+            LOGGER.info()
+                    .setMessage("Upserted alert")
+                    .addData("alert", alert)
+                    .addData("isCreated", created)
+                    .log();
+        } catch (final IOException e) {
+            LOGGER.error()
+                    .setMessage("Failed to close the transaction")
+                    .log();
+            throw new PersistenceException(e);
+            //CHECKSTYLE.OFF: IllegalCatchCheck
+        } catch (final RuntimeException e) {
+            //CHECKSTYLE.ON
+            LOGGER.error()
+                    .setMessage("Failed to upsert alert")
+                    .addData("alert", alert)
+                    .log();
+            throw new PersistenceException(e);
+        }
+    }
+
+    private void assertIsOpen() {
+        assertIsOpen(true);
+    }
+
+    private void assertIsOpen(final boolean expectedState) {
+        if (_isOpen.get() != expectedState) {
+            throw new IllegalStateException(String.format("Alert repository is not %s", expectedState ? "open" : "closed"));
+        }
+    }
+
+    private NagiosExtension getNagiosExtensionFromMap(final Map<String, Object> extensionsMap) {
+        if (extensionsMap == null || extensionsMap.isEmpty()) {
+            return null;
+        }
+        final NagiosExtension nagiosExtension = new NagiosExtension();
+        nagiosExtension.setSeverity((String) extensionsMap.get(NAGIOS_EXTENSION_SEVERITY_KEY));
+        nagiosExtension.setNotify((String) extensionsMap.get(NAGIOS_EXTENSION_NOTIFY_KEY));
+        nagiosExtension.setMaxCheckAttempts((Integer) extensionsMap.get(NAGIOS_EXTENSION_MAX_CHECK_ATTEMPTS_KEY));
+        nagiosExtension.setFreshnessThreshold((Integer) extensionsMap.get(NAGIOS_EXTENSION_FRESHNESS_THRESHOLD_KEY));
+        return nagiosExtension;
+    }
+
+    private Map<String, Object> mergeExtensions(final NagiosExtension nagiosExtension) {
```

Comment:
So to be clear this obfuscation of the extensions into a map should happen at the view layer. If this is a temporary state to ensure backwards compatibility that's fine. The existing v1 endpoints will continue to return a map data structure for extensions while you should also write v2 endpoints which return structured extensions (e.g. NagiosExtensions with named fields). The latter will facilitate construction of the user interface while the former retains compatibility with TsdAggregator/ClusterAggregation and Nagios which query this endpoint.